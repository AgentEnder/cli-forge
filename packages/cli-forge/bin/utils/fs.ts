import { existsSync, mkdirSync } from 'node:fs';

export function ensureDirSync(dir: string) {
  try {
    mkdirSync(dir, { recursive: true });
  } catch (e) {
    if (!existsSync(dir)) {
      throw new Error(`Could not create directory: ${dir}`, { cause: e });
    }
  }
}
