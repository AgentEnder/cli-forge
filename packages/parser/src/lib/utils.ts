export function fromDashedToCamelCase(str: string) {
  return str.replace(/-([a-z])/g, (match, letter) => letter.toUpperCase());
}

export function fromCamelOrDashedCaseToConstCase(str: string) {
  const parts = [];
  let currentPart = '';
  let prev: string | undefined;
  for (const char of str) {
    if (char === '-') {
      if (currentPart.length) {
        parts.push(currentPart);
      }
      currentPart = '';
      continue;
    }
    // Previous was lower case and current is upper case
    if (prev && prev?.toLowerCase() === prev && char === char.toUpperCase()) {
      parts.push(currentPart);
      currentPart = char;
      // Handle someCLITool -> SOME_CLI_TOOL
    } else if (
      prev?.toUpperCase() === prev &&
      char === char.toLowerCase() &&
      currentPart.length > 1
    ) {
      parts.push(currentPart.slice(0, -1));
      currentPart = currentPart.slice(-1) + char;
    } else {
      currentPart += char;
    }
    prev = char;
  }
  parts.push(currentPart);
  return parts.join('_').toUpperCase();
}

export function getEnvKey(prefix: string | undefined, key: string) {
  return prefix
    ? `${prefix}_${fromCamelOrDashedCaseToConstCase(key)}`
    : fromCamelOrDashedCaseToConstCase(key);
}
