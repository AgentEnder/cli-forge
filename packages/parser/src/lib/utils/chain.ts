// This file pains me quite a bit. Before any feedback is given, consider the following:
// - There is currently no way to infer the types of a `...rest` parameter individually.
// - This is how rxjs does it, and that's because its really the only way in current TS / JS
//   - See: https://github.com/ReactiveX/rxjs/blob/master/packages%2Fobservable%2Fsrc%2Fobservable.ts#L788
// - This isn't maintained or even written by hand, aside from the final implementation. The various
//   overload signatures are copilot generated.

import { OptionConfig } from '../option-types';
import { ArgvParser, ParsedArgs } from '../parser';

type UnaryFunction<T, R> = (arg: T) => R;

/**
 * Applies a series of functions to an initial value, passing the result of each function to the next.
 *
 * Used to convert code that looks like:
 * ```typescript
 * const a = addB(addC({ a: 'a' }));
 * ```
 * to:
 * ```typescript
 * const a = chain({ a: 'a' }, addB, addC);
 * ```
 *
 * See [composable-options](/examples/composable-options) for an example of how this can be used.
 *
 * @param initial Initial value to pass through the chain
 * @param fns Functions to apply to the initial value, and each subsequent value
 * @returns Updated value after all functions have been applied
 * @hidden docs for {@link chain} are found on the main signature.
 */
export function chain<T0, R>(initial: T0, fn0: UnaryFunction<T0, R>): R;

/**
 * Applies a series of functions to an initial value, passing the result of each function to the next.
 *
 * Used to convert code that looks like:
 * ```typescript
 * const a = addB(addC({ a: 'a' }));
 * ```
 * to:
 * ```typescript
 * const a = chain({ a: 'a' }, addB, addC);
 * ```
 *
 * See [composable-options](/examples/composable-options) for an example of how this can be used.
 *
 * @param initial Initial value to pass through the chain
 * @param fns Functions to apply to the initial value, and each subsequent value
 * @returns Updated value after all functions have been applied
 * @hidden docs for {@link chain} are found on the main signature.
 */
export function chain<T0, T1, R>(
  initial: T0,
  fn0: UnaryFunction<T0, T1>,
  fn1: UnaryFunction<T1, R>
): R;

/**
 * Applies a series of functions to an initial value, passing the result of each function to the next.
 *
 * Used to convert code that looks like:
 * ```typescript
 * const a = addB(addC({ a: 'a' }));
 * ```
 * to:
 * ```typescript
 * const a = chain({ a: 'a' }, addB, addC);
 * ```
 *
 * See [composable-options](/examples/composable-options) for an example of how this can be used.
 *
 * @param initial Initial value to pass through the chain
 * @param fns Functions to apply to the initial value, and each subsequent value
 * @returns Updated value after all functions have been applied
 * @hidden docs for {@link chain} are found on the main signature.
 */
export function chain<T0, T1, T2, R>(
  initial: T0,
  fn0: UnaryFunction<T0, T1>,
  fn1: UnaryFunction<T1, T2>,
  fn2: UnaryFunction<T2, R>
): R;

/**
 * Applies a series of functions to an initial value, passing the result of each function to the next.
 *
 * Used to convert code that looks like:
 * ```typescript
 * const a = addB(addC({ a: 'a' }));
 * ```
 * to:
 * ```typescript
 * const a = chain({ a: 'a' }, addB, addC);
 * ```
 *
 * See [composable-options](/examples/composable-options) for an example of how this can be used.
 *
 * @param initial Initial value to pass through the chain
 * @param fns Functions to apply to the initial value, and each subsequent value
 * @returns Updated value after all functions have been applied
 * @hidden docs for {@link chain} are found on the main signature.
 */
export function chain<T0, T1, T2, T3, R>(
  initial: T0,
  fn0: UnaryFunction<T0, T1>,
  fn1: UnaryFunction<T1, T2>,
  fn2: UnaryFunction<T2, T3>,
  fn3: UnaryFunction<T3, R>
): R;

/**
 * Applies a series of functions to an initial value, passing the result of each function to the next.
 *
 * Used to convert code that looks like:
 * ```typescript
 * const a = addB(addC({ a: 'a' }));
 * ```
 * to:
 * ```typescript
 * const a = chain({ a: 'a' }, addB, addC);
 * ```
 *
 * See [composable-options](/examples/composable-options) for an example of how this can be used.
 *
 * @param initial Initial value to pass through the chain
 * @param fns Functions to apply to the initial value, and each subsequent value
 * @returns Updated value after all functions have been applied
 * @hidden docs for {@link chain} are found on the main signature.
 */
export function chain<T0, T1, T2, T3, T4, R>(
  initial: T0,
  fn0: UnaryFunction<T0, T1>,
  fn1: UnaryFunction<T1, T2>,
  fn2: UnaryFunction<T2, T3>,
  fn3: UnaryFunction<T3, T4>,
  fn4: UnaryFunction<T4, R>
): R;

/**
 * Applies a series of functions to an initial value, passing the result of each function to the next.
 *
 * Used to convert code that looks like:
 * ```typescript
 * const a = addB(addC({ a: 'a' }));
 * ```
 * to:
 * ```typescript
 * const a = chain({ a: 'a' }, addB, addC);
 * ```
 *
 * See [composable-options](/examples/composable-options) for an example of how this can be used.
 *
 * @param initial Initial value to pass through the chain
 * @param fns Functions to apply to the initial value, and each subsequent value
 * @returns Updated value after all functions have been applied
 * @hidden docs for {@link chain} are found on the main signature.
 */
export function chain<T0, T1, T2, T3, T4, T5, R>(
  initial: T0,
  fn0: UnaryFunction<T0, T1>,
  fn1: UnaryFunction<T1, T2>,
  fn2: UnaryFunction<T2, T3>,
  fn3: UnaryFunction<T3, T4>,
  fn4: UnaryFunction<T4, T5>,
  fn5: UnaryFunction<T5, R>
): R;

/**
 * Applies a series of functions to an initial value, passing the result of each function to the next.
 *
 * Used to convert code that looks like:
 * ```typescript
 * const a = addB(addC({ a: 'a' }));
 * ```
 * to:
 * ```typescript
 * const a = chain({ a: 'a' }, addB, addC);
 * ```
 *
 * See [composable-options](/examples/composable-options) for an example of how this can be used.
 *
 * @param initial Initial value to pass through the chain
 * @param fns Functions to apply to the initial value, and each subsequent value
 * @returns Updated value after all functions have been applied
 * @hidden docs for {@link chain} are found on the main signature.
 */
export function chain<T0, T1, T2, T3, T4, T5, T6, R>(
  initial: T0,
  fn0: UnaryFunction<T0, T1>,
  fn1: UnaryFunction<T1, T2>,
  fn2: UnaryFunction<T2, T3>,
  fn3: UnaryFunction<T3, T4>,
  fn4: UnaryFunction<T4, T5>,
  fn5: UnaryFunction<T5, T6>,
  fn6: UnaryFunction<T6, R>
): R;

/**
 * Applies a series of functions to an initial value, passing the result of each function to the next.
 *
 * Used to convert code that looks like:
 * ```typescript
 * const a = addB(addC({ a: 'a' }));
 * ```
 * to:
 * ```typescript
 * const a = chain({ a: 'a' }, addB, addC);
 * ```
 *
 * See [composable-options](/examples/composable-options) for an example of how this can be used.
 *
 * @param initial Initial value to pass through the chain
 * @param fns Functions to apply to the initial value, and each subsequent value
 * @returns Updated value after all functions have been applied
 * @hidden docs for {@link chain} are found on the main signature.
 */
export function chain<T0, T1, T2, T3, T4, T5, T6, T7, R>(
  initial: T0,
  fn0: UnaryFunction<T0, T1>,
  fn1: UnaryFunction<T1, T2>,
  fn2: UnaryFunction<T2, T3>,
  fn3: UnaryFunction<T3, T4>,
  fn4: UnaryFunction<T4, T5>,
  fn5: UnaryFunction<T5, T6>,
  fn6: UnaryFunction<T6, T7>,
  fn7: UnaryFunction<T7, R>
): R;

/**
 * Applies a series of functions to an initial value, passing the result of each function to the next.
 *
 * Used to convert code that looks like:
 * ```typescript
 * const a = addB(addC({ a: 'a' }));
 * ```
 * to:
 * ```typescript
 * const a = chain({ a: 'a' }, addB, addC);
 * ```
 *
 * See [composable-options](/examples/composable-options) for an example of how this can be used.
 *
 * @param initial Initial value to pass through the chain
 * @param fns Functions to apply to the initial value, and each subsequent value
 * @returns Updated value after all functions have been applied
 * @hidden docs for {@link chain} are found on the main signature.
 */
export function chain<T0, T1, T2, T3, T4, T5, T6, T7, T8, R>(
  initial: T0,
  fn0: UnaryFunction<T0, T1>,
  fn1: UnaryFunction<T1, T2>,
  fn2: UnaryFunction<T2, T3>,
  fn3: UnaryFunction<T3, T4>,
  fn4: UnaryFunction<T4, T5>,
  fn5: UnaryFunction<T5, T6>,
  fn6: UnaryFunction<T6, T7>,
  fn7: UnaryFunction<T7, T8>,
  fn8: UnaryFunction<T8, R>
): R;

/**
 * Applies a series of functions to an initial value, passing the result of each function to the next.
 *
 * Used to convert code that looks like:
 * ```typescript
 * const a = addB(addC({ a: 'a' }));
 * ```
 * to:
 * ```typescript
 * const a = chain({ a: 'a' }, addB, addC);
 * ```
 *
 * See [composable-options](/examples/composable-options) for an example of how this can be used.
 *
 * @param initial Initial value to pass through the chain
 * @param fns Functions to apply to the initial value, and each subsequent value
 * @returns Updated value after all functions have been applied
 * @hidden docs for {@link chain} are found on the main signature.
 */
export function chain<T0, T1, T2, T3, T4, T5, T6, T7, T8, T9, R>(
  initial: T0,
  fn0: UnaryFunction<T0, T1>,
  fn1: UnaryFunction<T1, T2>,
  fn2: UnaryFunction<T2, T3>,
  fn3: UnaryFunction<T3, T4>,
  fn4: UnaryFunction<T4, T5>,
  fn5: UnaryFunction<T5, T6>,
  fn6: UnaryFunction<T6, T7>,
  fn7: UnaryFunction<T7, T8>,
  fn8: UnaryFunction<T8, T9>,
  fn9: UnaryFunction<T9, R>
): R;

/**
 * Applies a series of functions to an initial value, passing the result of each function to the next.
 *
 * Used to convert code that looks like:
 * ```typescript
 * const a = addB(addC({ a: 'a' }));
 * ```
 * to:
 * ```typescript
 * const a = chain({ a: 'a' }, addB, addC);
 * ```
 *
 * See [composable-options](/examples/composable-options) for an example of how this can be used.
 *
 * @param initial Initial value to pass through the chain
 * @param fns Functions to apply to the initial value, and each subsequent value
 * @returns Updated value after all functions have been applied
 * @hidden docs for {@link chain} are found on the main signature.
 */
export function chain<T0, T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, R>(
  initial: T0,
  fn0: UnaryFunction<T0, T1>,
  fn1: UnaryFunction<T1, T2>,
  fn2: UnaryFunction<T2, T3>,
  fn3: UnaryFunction<T3, T4>,
  fn4: UnaryFunction<T4, T5>,
  fn5: UnaryFunction<T5, T6>,
  fn6: UnaryFunction<T6, T7>,
  fn7: UnaryFunction<T7, T8>,
  fn8: UnaryFunction<T8, T9>,
  fn9: UnaryFunction<T9, T10>,
  fn10: UnaryFunction<T10, R>
): R;

/**
 * Applies a series of functions to an initial value, passing the result of each function to the next.
 *
 * Used to convert code that looks like:
 * ```typescript
 * const a = addB(addC({ a: 'a' }));
 * ```
 * to:
 * ```typescript
 * const a = chain({ a: 'a' }, addB, addC);
 * ```
 *
 * See [composable-options](/examples/composable-options) for an example of how this can be used.
 *
 * @param initial Initial value to pass through the chain
 * @param fns Functions to apply to the initial value, and each subsequent value
 * @returns Updated value after all functions have been applied
 * @hidden docs for {@link chain} are found on the main signature.
 */
export function chain<T0, T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, R>(
  initial: T0,
  fn0: UnaryFunction<T0, T1>,
  fn1: UnaryFunction<T1, T2>,
  fn2: UnaryFunction<T2, T3>,
  fn3: UnaryFunction<T3, T4>,
  fn4: UnaryFunction<T4, T5>,
  fn5: UnaryFunction<T5, T6>,
  fn6: UnaryFunction<T6, T7>,
  fn7: UnaryFunction<T7, T8>,
  fn8: UnaryFunction<T8, T9>,
  fn9: UnaryFunction<T9, T10>,
  fn10: UnaryFunction<T10, T11>,
  fn11: UnaryFunction<T11, R>
): R;

/**
 * Applies a series of functions to an initial value, passing the result of each function to the next.
 *
 * Used to convert code that looks like:
 * ```typescript
 * const a = addB(addC({ a: 'a' }));
 * ```
 * to:
 * ```typescript
 * const a = chain({ a: 'a' }, addB, addC);
 * ```
 *
 * See [composable-options](/examples/composable-options) for an example of how this can be used.
 *
 * @param initial Initial value to pass through the chain
 * @param fns Functions to apply to the initial value, and each subsequent value
 * @returns Updated value after all functions have been applied
 * @hidden docs for {@link chain} are found on the main signature.
 */
export function chain<T0, T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, T12, R>(
  initial: T0,
  fn0: UnaryFunction<T0, T1>,
  fn1: UnaryFunction<T1, T2>,
  fn2: UnaryFunction<T2, T3>,
  fn3: UnaryFunction<T3, T4>,
  fn4: UnaryFunction<T4, T5>,
  fn5: UnaryFunction<T5, T6>,
  fn6: UnaryFunction<T6, T7>,
  fn7: UnaryFunction<T7, T8>,
  fn8: UnaryFunction<T8, T9>,
  fn9: UnaryFunction<T9, T10>,
  fn10: UnaryFunction<T10, T11>,
  fn11: UnaryFunction<T11, T12>,
  fn12: UnaryFunction<T12, R>
): R;

/**
 * Applies a series of functions to an initial value, passing the result of each function to the next.
 *
 * Used to convert code that looks like:
 * ```typescript
 * const a = addB(addC({ a: 'a' }));
 * ```
 * to:
 * ```typescript
 * const a = chain({ a: 'a' }, addB, addC);
 * ```
 *
 * See [composable-options](/examples/composable-options) for an example of how this can be used.
 *
 * @param initial Initial value to pass through the chain
 * @param fns Functions to apply to the initial value, and each subsequent value
 * @returns Updated value after all functions have been applied
 * @hidden docs for {@link chain} are found on the main signature.
 */
export function chain<
  T0,
  T1,
  T2,
  T3,
  T4,
  T5,
  T6,
  T7,
  T8,
  T9,
  T10,
  T11,
  T12,
  T13,
  R
>(
  initial: T0,
  fn0: UnaryFunction<T0, T1>,
  fn1: UnaryFunction<T1, T2>,
  fn2: UnaryFunction<T2, T3>,
  fn3: UnaryFunction<T3, T4>,
  fn4: UnaryFunction<T4, T5>,
  fn5: UnaryFunction<T5, T6>,
  fn6: UnaryFunction<T6, T7>,
  fn7: UnaryFunction<T7, T8>,
  fn8: UnaryFunction<T8, T9>,
  fn9: UnaryFunction<T9, T10>,
  fn10: UnaryFunction<T10, T11>,
  fn11: UnaryFunction<T11, T12>,
  fn12: UnaryFunction<T12, T13>,
  fn13: UnaryFunction<T13, R>
): R;

/**
 * Applies a series of functions to an initial value, passing the result of each function to the next.
 *
 * Used to convert code that looks like:
 * ```typescript
 * const a = addB(addC({ a: 'a' }));
 * ```
 * to:
 * ```typescript
 * const a = chain({ a: 'a' }, addB, addC);
 * ```
 *
 * See [composable-options](/examples/composable-options) for an example of how this can be used.
 *
 * @param initial Initial value to pass through the chain
 * @param fns Functions to apply to the initial value, and each subsequent value
 * @returns Updated value after all functions have been applied
 * @hidden docs for {@link chain} are found on the main signature.
 */
export function chain<
  T0,
  T1,
  T2,
  T3,
  T4,
  T5,
  T6,
  T7,
  T8,
  T9,
  T10,
  T11,
  T12,
  T13,
  T14,
  R
>(
  initial: T0,
  fn0: UnaryFunction<T0, T1>,
  fn1: UnaryFunction<T1, T2>,
  fn2: UnaryFunction<T2, T3>,
  fn3: UnaryFunction<T3, T4>,
  fn4: UnaryFunction<T4, T5>,
  fn5: UnaryFunction<T5, T6>,
  fn6: UnaryFunction<T6, T7>,
  fn7: UnaryFunction<T7, T8>,
  fn8: UnaryFunction<T8, T9>,
  fn9: UnaryFunction<T9, T10>,
  fn10: UnaryFunction<T10, T11>,
  fn11: UnaryFunction<T11, T12>,
  fn12: UnaryFunction<T12, T13>,
  fn13: UnaryFunction<T13, T14>,
  fn14: UnaryFunction<T14, R>
): R;

/**
 * Applies a series of functions to an initial value, passing the result of each function to the next.
 *
 * Used to convert code that looks like:
 * ```typescript
 * const a = addB(addC({ a: 'a' }));
 * ```
 * to:
 * ```typescript
 * const a = chain({ a: 'a' }, addB, addC);
 * ```
 *
 * See [composable-options](/examples/composable-options) for an example of how this can be used.
 *
 * @param {T0} initial Initial value to pass through the chain
 * @param fns Functions to apply to the initial value, and each subsequent value
 * @returns {R} Updated value after all functions have been applied
 */
export function chain<
  T0,
  T1,
  T2,
  T3,
  T4,
  T5,
  T6,
  T7,
  T8,
  T9,
  T10,
  T11,
  T12,
  T13,
  T14,
  T15,
  R
>(
  initial: T0,
  fn0: UnaryFunction<T0, T1>,
  fn1: UnaryFunction<T1, T2>,
  fn2: UnaryFunction<T2, T3>,
  fn3: UnaryFunction<T3, T4>,
  fn4: UnaryFunction<T4, T5>,
  fn5: UnaryFunction<T5, T6>,
  fn6: UnaryFunction<T6, T7>,
  fn7: UnaryFunction<T7, T8>,
  fn8: UnaryFunction<T8, T9>,
  fn9: UnaryFunction<T9, T10>,
  fn10: UnaryFunction<T10, T11>,
  fn11: UnaryFunction<T11, T12>,
  fn12: UnaryFunction<T12, T13>,
  fn13: UnaryFunction<T13, T14>,
  fn14: UnaryFunction<T14, T15>,
  fn15: UnaryFunction<T15, R>
): R;

/**
 * Applies a series of functions to an initial value, passing the result of each function to the next.
 *
 * Used to convert code that looks like:
 * ```typescript
 * const a = addB(addC({ a: 'a' }));
 * ```
 * to:
 * ```typescript
 * const a = chain({ a: 'a' }, addB, addC);
 * ```
 *
 * See [composable-options](/examples/composable-options) for an example of how this can be used.
 *
 * NOTE: Seeing a return type of `unknown`? This is likely due to the number of functions passed to
 * `chain` exceeding the number of overloads provided. There is a technical limitation in TypeScript
 * that means we need to provide a finite number of overloads. While we anticipate that 16 overloads
 * should be enough for most use cases, if you need more you can combine chains together like below:
 * ```typescript
 * const a = chain(
 *    { a: 'a' },
 *   addB,
 *   (iv) => chain(iv, addC, addD, addE),
 * );
 * ```
 *
 * @param initial Initial value to pass through the chain
 * @param fns Functions to apply to the initial value, and each subsequent value
 * @returns Updated value after all functions have been applied
 */
export function chain<T0>(
  initial: T0,
  ...fns: UnaryFunction<any, any>[]
): unknown {
  return fns.reduce((acc, fn) => fn(acc), initial);
}

/**
 * A composition helper to be used with {@link chain}.
 *
 * @param name The name of the option to add
 * @param config The configuration for the option
 * @returns A unary function that adds an option to an {@link ArgvParser}.
 */
export function makeComposableOption<
  const TKey extends string,
  const TOptionConfig extends OptionConfig
>(name: TKey, config: TOptionConfig) {
  return <const T extends ParsedArgs>(argv: ArgvParser<T>) =>
    argv.option(name, config);
}
