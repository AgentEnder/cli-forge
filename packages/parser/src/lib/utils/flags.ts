import { fromDashedToCamelCase } from './case-transformations';

export function isFlag(str: string): str is `-${string}` {
  return str.startsWith('-');
}
export function readArgKeys(str: `-${string}`): string[] {
  // Long flags (e.g. --foo)
  if (str.startsWith('--')) {
    const key = str.slice(2);
    if (key.includes('-')) {
      return [fromDashedToCamelCase(key)];
    }
    return [str.slice(2)];
    // Short flag combinations (e.g. -xvf)
  } else if (str.startsWith('-')) {
    return str.slice(1).split('');
  }
  throw new Error(`Invalid flag ${str}`);
}
export function isNextFlag(str: string) {
  return str.startsWith('--') || str.startsWith('-');
}
