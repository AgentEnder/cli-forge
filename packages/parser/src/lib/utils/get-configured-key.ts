import { OptionConfig } from '../option-types';
import { ParsedArgs } from '../parser';

export function getConfiguredOptionKey<T extends ParsedArgs>(
  key: string,
  configuredOptions: Partial<Record<keyof T, OptionConfig>>
): keyof T | undefined {
  if (key in configuredOptions) {
    return key as keyof T;
  }

  function normalizeNegatedBooleanKey(key: string) {
    if (
      key.startsWith('no') &&
      key.length > 2 &&
      key[2] === key[2].toUpperCase()
    ) {
      const stripped = key.slice(2);
      return [stripped[0].toLowerCase(), stripped.slice(1)].join('');
    }
    return key;
  }

  function normalizeObjectOptionKey(key: string) {
    if (key.includes('.')) {
      return key.split('.')[0];
    }
    return key;
  }

  // Handles booleans passed as `--no-foo`
  const normalizedKey = normalizeObjectOptionKey(
    normalizeNegatedBooleanKey(key)
  );

  if (normalizedKey in configuredOptions) {
    return normalizedKey as keyof T;
  }

  for (const configuredKey in configuredOptions) {
    const config = configuredOptions[configuredKey];
    if (config?.alias?.includes(key)) {
      return configuredKey as keyof T;
    }
    // Handles negated booleans that are aliased
    if (config?.type === 'boolean' && config.alias?.includes(normalizedKey)) {
      return configuredKey as keyof T;
    }
  }
  return undefined;
}
