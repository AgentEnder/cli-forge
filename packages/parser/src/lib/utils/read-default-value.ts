import {
  Default,
  OptionConfig,
  PlainDefaultValue,
  OptionConfigToType,
} from '../option-types';

export function readDefaultValue<const T extends OptionConfig>(
  option: T
): [OptionConfigToType<T> | undefined, string | undefined] {
  if (option.default !== undefined) {
    const declaredDefault = option.default as any as Default<T>;
    if (typeof declaredDefault === 'object') {
      if ('value' in declaredDefault) {
        const value = declaredDefault.value as any as OptionConfigToType<T>;
        return [value, declaredDefault.description];
      }
      if ('factory' in declaredDefault) {
        const factory =
          declaredDefault.factory as any as () => OptionConfigToType<T>;
        return [factory(), declaredDefault.description];
      }
    } else {
      return [declaredDefault, undefined];
    }
  }
  return [undefined, undefined];
}
