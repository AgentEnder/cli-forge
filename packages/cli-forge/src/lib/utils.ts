import { existsSync } from 'fs';
import { join, resolve } from 'path';

export function getCallingFile() {
  // Since this function lives in a utility file, the parent file
  // would just be the file that invokes getCallingFile... We actually
  // want the parent of the file that calls this, which is this file's
  // grandparent.
  let grandparentFile: string | undefined;

  // Overrides prepare stack trace to be an identity fn, to get the callsites
  // associated with the current error.
  const _pst = Error.prepareStackTrace;
  Error.prepareStackTrace = function (err, stack) {
    return stack;
  };

  try {
    const err = new Error();
    const callsites = err.stack as any as NodeJS.CallSite[];

    let currentfile = callsites.shift()!.getFileName();
    let parentfile: string | undefined;

    while (callsites.length) {
      const callerfile = callsites.shift()!.getFileName();

      // We've reached the parent file
      if (currentfile !== callerfile) {
        // We've reached the grandparent file
        if (parentfile && parentfile !== callerfile) {
          grandparentFile = parentfile;
          break;
        }
        parentfile = callerfile;
      }
    }
  } finally {
    Error.prepareStackTrace = _pst;
  }

  return grandparentFile;
}

export function getParentPackageJson(searchPath: string) {
  let currentPath = searchPath;
  let packageJsonPath: string | undefined;

  while (true) {
    const packagePath = join(currentPath, 'package.json');

    if (existsSync(packagePath)) {
      packageJsonPath = packagePath;
      break;
    }

    const nextPath = resolve(currentPath, '..');

    if (nextPath === currentPath) {
      break;
    }

    currentPath = nextPath;
  }

  if (!packageJsonPath) {
    throw new Error('Could not find package.json');
  }

  return require(packageJsonPath) as {
    name: string;
    version: string;
    bin?: {
      [cmd: string]: string;
    };
    dependencies?: Record<string, string>;
  };
}
