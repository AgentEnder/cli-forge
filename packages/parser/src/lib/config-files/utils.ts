import { existsSync } from 'node:fs';
import { dirname, join } from 'node:path';

export function traverseForFile(
  configFileName: string,
  searchDir = process.cwd()
) {
  // Traverse up file system to find json file
  let prev: string | undefined;
  let current = searchDir;
  while (prev !== current) {
    prev = current;
    const testPath = join(current, configFileName);
    if (existsSync(testPath)) {
      return testPath;
    } else {
      current = dirname(current);
    }
  }
  return undefined;
}
