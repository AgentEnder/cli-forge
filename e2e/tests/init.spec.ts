import { join } from 'node:path';

import { e2eSubDir, ensureCleanWorkingDirectory } from '../utils/setup';
import { runCommand } from '../utils/child_process';
import { checkFilesExist } from '../utils/fs';

describe('init', () => {
  beforeEach(() => {
    ensureCleanWorkingDirectory();
  });

  describe.each([['ts'], ['js']])(`--format %s`, (format) => {
    it('should generate a new CLI', async () => {
      await runCommand(
        'npx cli-forge@e2e init my-cli --format ' + format,
        [],
        {}
      );
      expect(() =>
        checkFilesExist(
          ['package.json', join('bin', 'my-cli.' + format)].map((f) =>
            join(e2eSubDir, f)
          )
        )
      ).not.toThrow();

      let { stdout } = await runCommand(
        'npx -y tsx ./bin/my-cli hello world',
        [],
        {}
      );
      expect(stdout).toMatchSnapshot('command output');

      ({ stdout } = await runCommand('npx -y tsx ./bin/my-cli --help', [], {}));
      expect(stdout).toMatchSnapshot('help text');

      ({ stdout } = await runCommand(
        'npx -y tsx ./bin/my-cli hello --help',
        [],
        {}
      ));
      expect(stdout).toMatchSnapshot('subcommand help text');

      //
      if (format === 'js') {
        await runCommand(
          'npx cli-forge generate-documentation ./bin/my-cli',
          [],
          {}
        );
      } else {
        await runCommand(
          'npx tsx ./node_modules/.bin/cli-forge generate-documentation ./bin/my-cli',
          [],
          {}
        );
      }

      expect(() =>
        checkFilesExist(
          [
            'docs',
            join('docs', 'my-cli', 'index.md'),
            join('docs', 'my-cli', 'hello.md'),
          ].map((f) => join(e2eSubDir, f))
        )
      ).not.toThrow();
    });
  });
});
