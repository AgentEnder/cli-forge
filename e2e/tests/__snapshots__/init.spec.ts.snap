// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`init --format js should generate a new CLI: command output 1`] = `
"hello world
"
`;

exports[`init --format js should generate a new CLI: help text 1`] = `
"Usage: my-cli

Commands:
  hello

Options:
  --help - Show help for the current command
 
Run \`my-cli [command] --help\` for more information on a command
"
`;

exports[`init --format js should generate a new CLI: subcommand help text 1`] = `
"Usage: my-cli hello [name]

Options:
  --help - Show help for the current command
"
`;

exports[`init --format ts should generate a new CLI: command output 1`] = `
"hello world
"
`;

exports[`init --format ts should generate a new CLI: help text 1`] = `
"Usage: my-cli

Commands:
  hello

Options:
  --help - Show help for the current command
 
Run \`my-cli [command] --help\` for more information on a command
"
`;

exports[`init --format ts should generate a new CLI: subcommand help text 1`] = `
"Usage: my-cli hello [name]

Options:
  --help - Show help for the current command
"
`;
